capsules/extra/src/can.rs

```diff
@@ -443,80 +445,64 @@ impl<'a, Can: can::Can> can::ReceiveClient<{ can::STANDARD_CAN_PACKET_SIZE }>
-                                                if app_data.receive_index + len > user_buffer.len()
-                                                {
-                                                    app_data.lost_messages += 1;
-                                                    Err(ErrorCode::SIZE)
-                                                } else {
-                                                    let r = user_buffer[app_data.receive_index
-                                                        ..app_data.receive_index + len]
-                                                        .copy_from_slice_or_err(&buffer[0..len]);
-                                                    if r.is_ok() {
-                                                        app_data.receive_index += len;
-                                                    }
-                                                    r
-                                                }
-                                            })
-                                            .unwrap_or_else(|err| err.into())
-                                    },
-                                )
-                        })
-                        .unwrap_or_else(|err| err.into())
-                }) {
+                let res: Result<StreamingProcessSliceMeta, ErrorCode> =
+                    self.processid.map_or(Err(ErrorCode::NOMEM), |processid| {
+                        self.processes
+                            .enter(processid, |app_data, kernel_data| {
+                                kernel_data
+                                    .get_readwrite_processbuffer(rw_allow::RW_ALLOW_BUFFER)
+                                    .map_or_else(
+                                        |err| Err(err.into()),
+                                        |buffer_ref| {
+                                            buffer_ref
+                                                .mut_enter(|user_slice| {
+                                                    // ???
+                                                    //
+                                                    // shared_len = user_buffer.offset()?;
+                                                    // // This uses the ringbuffer interpretation of the
+                                                    // // ProcessBufferSlice
+                                                    // new_buffer = match user_buffer.offset() {
+                                                    //     Err(ErrorCode::INVAL) => {
+                                                    //         user_buffer.reset().map(|()| 0)
+                                                    //     }
+                                                    //     value => value,
+                                                    // }? == 0;
+                                                    //
```

Comment:
?

Reply:
Yeah, exactly.

I don't really understand these changes as done in #4023, but @alexandruradovici wants to have a look at this in the next few days.

Reply:
This part was trying to fix an invalid buffer shared by the app. The original proposal had a verification byte (xor). The `offset` function would fail if the verification byte would be wrongly set. In this case, the can driver resets the buffer by correctly writing the offset `0` and setting the verification byte to the correct value.

In @lschuermann implementation, I don't think that this is necessary.